Added a live preview for the statemachine example

emilkrebs opened the pull request:
![State Machine](https://user-images.githubusercontent.com/68400102/222200220-2d9aab68-effa-4385-b5a8-e81bb7f67dd3.png)

**What does it do?**
The preview basically renders what you've written with the DSL. 

Click the event buttons to change to the state defined in your code or click on a state to change it directly.

Everything your change in the code will automatically be rendered - if there are any diagnostics (exceptions) in your code it will fail and show them for you.

**How is it made?**

The whole preview was made with react using the `@typefox/monaco-editor-react` package (https://www.npmjs.com/package/@typefox/monaco-editor-react).

When receiving the `onDocumentChange` notification it [decodes the AST](https://github.com/langium/langium-website/blob/langium-preview-develop/hugo/assets/scripts/ast/ast-tools.ts) and visualizes it.

Have a look at this file: https://github.com/langium/langium-website/blob/langium-preview-develop/hugo/assets/scripts/statemachine.tsx everything happens there.

emilkrebs commented:
I had some issues while rebasing, please check if I did change something that wasn't intended to.

Lotes commented on hugo/layouts/partials/langium-scripts.html:
```diff
@@ -1,4 +1,4 @@
-<script>
+{{/*  <script>
     // try {
```
Commented-out code should be removed... it is an indicator that the code is not finished yet. Normally when it is production ready you should be free of it.

Lotes commented on hugo/assets/scripts/statemachine.tsx:
```diff
@@ -0,0 +1,356 @@
+              workerType="classic"
+              languageId="statemachine"
+              text={`// Create your own statemachine here!
+statemachine TrafficLight
```
Move that giant string to a own file or at least a variable. I disturbs a bit while reading the code.

Lotes commented on hugo/assets/scripts/statemachine.tsx:
```diff
@@ -0,0 +1,356 @@
+import {
```
This file is very long, you should split it somehow to smaller chucks. Maybe one per concept/class

Lotes commented on hugo/assets/scripts/domainmodel.tsx:
```diff
@@ -0,0 +1,116 @@
+      <div className="w-full h-full border border-emeraldLangium justify-center self-center flex">
+        <div className="float-left w-1/2 h-full border-r border-emeraldLangium">
+          <div className="wrapper relative bg-white dark:bg-gray-900" >
+            <MonacoEditorReactComp ref={this.monacoEditor} onTextChanged={this.onTextChanged} webworkerUri="../showcase/libs/worker/domainmodelServerWorker.js" workerName='LS' workerType='classic' languageId="domainmodel" text={`// Define datatypes
```
Move giant strings to a own file with that constants. Improves readability

montymxb replied:
Can drop the `onTextChanged` callback here as well

Lotes commented on hugo/assets/scripts/domainmodel.tsx:
```diff
@@ -0,0 +1,116 @@
+);
+addMonacoStyles('monaco-editor-styles');
+
+const syntaxHighlighting = {
```
Also move this constant to its own file.

Lotes commented on hugo/assets/scripts/ast/statemachine.ts:
```diff
@@ -0,0 +1,145 @@
+        this.currentState = this.getInitialState(ast);
+    }
+
+    /**
```
Comments are actually very nice, but not when they reflect the name of the function.
You should talk about why is this piece of function like this and not what is obvious

montymxb replied:
Since this is statemachine specific, you could add that it's the starting state for the statemachine, based on the AST.

Lotes commented on hugo/assets/scripts/arithmetics.tsx:
```diff
@@ -0,0 +1,99 @@
+addMonacoStyles('monaco-editor-styles');
+ 
+const syntaxHighlighting = {
+    keywords: [
```
Move that constants to an own file, then the reader can focus better on the rest. Keep in mind that seeing such a big thing requires the reader to switch his mind to a different syntax/concept, which can be disturbing

Lotes commented on hugo/assets/scripts/arithmetics.tsx:
```diff
@@ -0,0 +1,99 @@
+} as monaco.languages.IMonarchLanguage;
+
+
+class App extends React.Component<{}, { hasError }> {
```
I think you are using React in a good way. I would recommend to use interfaces for generic type parameters. It happens pretty fast that you have to reuse them.

Lotes commented on hugo/assets/scripts/ast/ast-tools.ts:
```diff
@@ -0,0 +1,152 @@
+                const arrayIndex = parseInt(currentValue.substring(propertyIndex + 1));
+                // find array with prop & return via index
+                const array = (previousValue as unknown as Record<string, AstNode[]>)[property];
+                return array?.[arrayIndex];
```
Cool I learned something new :D

msujew reviewed (commented):
> I had some issues while rebasing, please check if I changed something that wasn't intended to.

It seems like you didn't rebase yet. Your branch doesn't contain the changes of the last five months.

By the way, the playground broke with your changes.

emilkrebs commented:
> > I had some issues while rebasing, please check if I changed something that wasn't intended to.
> 
> It seems like you didn't rebase yet. Your branch doesn't contain the changes of the last five months.
> 
> By the way, the playground broke with your changes.
> 
> On a general level, the change looks awesome! Good job 👍 

Yes, I reverted the rebate because it broke everything. When I have some time I will try again.

montymxb commented on hugo/assets/scripts/arithmetics.tsx:
```diff
@@ -0,0 +1,99 @@
+    };
+  }
+
+  onTextChanged(text: string, isDirty: boolean) {
```
Old function that we can remove

montymxb commented on hugo/assets/scripts/arithmetics.tsx:
```diff
@@ -0,0 +1,99 @@
+  monacoEditor: React.RefObject<MonacoEditorReactComp>;
+  constructor(props) {
+    super(props);
+    this.onTextChanged = this.onTextChanged.bind(this);
```
Can remove here as well, unused function below

montymxb commented on hugo/assets/scripts/arithmetics.tsx:
```diff
@@ -0,0 +1,99 @@
+      <div className="w-full h-full border border-emeraldLangium justify-center self-center flex">
+        <div className="float-left w-1/2 h-full border-r border-emeraldLangium">
+          <div className="wrapper relative bg-white dark:bg-gray-900" >
+            <MonacoEditorReactComp ref={this.monacoEditor} onTextChanged={this.onTextChanged} webworkerUri="../showcase/libs/worker/arithmeticsServerWorker.js" workerName='LS' workerType='classic' languageId="arithmetics" text={`Module example1
```
Can remove the `onTextChanged` callback

montymxb commented on hugo/assets/scripts/ast/ast-tools.ts:
```diff
@@ -0,0 +1,152 @@
+
+export class LangiumAST {
+    constructor() {
```
Can drop empty constructor

montymxb commented on hugo/assets/scripts/ast/ast-tools.ts:
```diff
@@ -0,0 +1,152 @@
+
+export class LangiumAST {
```
Would be good to leave a brief summary about the purpose/limitation of this class. Something akin to this is used to help with de-serializing & processing Langium AST nodes.

montymxb commented on hugo/assets/scripts/ast/statemachine.ts:
```diff
@@ -0,0 +1,145 @@
+import { type } from "os";
+import { AstNode, LangiumAST } from "./ast-tools";
```
We can drop the `type` and `LangiumAST` imports here, as they're unused.

montymxb commented on hugo/assets/scripts/ast/statemachine.ts:
```diff
@@ -0,0 +1,145 @@
+     * @param ast The AST to get the initial state from
+     * @returns 
+     */
+    private getInitialState(ast: StateMachineAstNode): StateMachineState {
```
Can probably drop the parameter, and instead reference `this.ast`, as it's set in the constructor above.

montymxb commented on hugo/assets/scripts/ast/statemachine.ts:
```diff
@@ -0,0 +1,145 @@
+
+
+
+export class StateMachineTools {
```
I quite like this one, nice abstraction on the statemachine 🙂

montymxb commented on hugo/assets/scripts/domainmodel.tsx:
```diff
@@ -0,0 +1,116 @@
+    };
+  }
+
+  onTextChanged(text: string, isDirty: boolean) {
```
Same here, must have been a vestige from before. Can drop this function, its binding above, and its usage below.

montymxb commented on hugo/assets/scripts/domainmodel.tsx:
```diff
@@ -0,0 +1,116 @@
+  monacoEditor: React.RefObject<MonacoEditorReactComp>;
+  constructor(props) {
+    super(props);
+    this.onTextChanged = this.onTextChanged.bind(this);
```
Unused binding

montymxb commented on hugo/assets/scripts/domainmodel.tsx:
```diff
@@ -0,0 +1,116 @@
+const root = createRoot(document.getElementById("root") as HTMLElement);
+root.render(
+  <App />
+); 
```
Always good to add a newline at the end 😉

montymxb commented on hugo/assets/scripts/statemachine.tsx:
```diff
@@ -0,0 +1,356 @@
+              return (
+                <Event
+                  isEnabled={statemachineTools.isEventEnabled(event)}
+                  handleClick={() => changeStates(statemachineTools.getNextState(event)!)}
```
I would be a bit careful with using non-null assertion op so frequently (`!`), as it can hide some issues down the road. Doesn't necessarily need to be addressed here however.

montymxb commented on hugo/babel.config.js:
```diff
@@ -0,0 +1,12 @@
+      presets,
+      plugins
+    };
+  }
```
New line would be good here

montymxb commented on hugo/content/showcase/domainmodel.html:
```diff
@@ -5,10 +5,9 @@
-<h3 style='color:white'>TODO: Implement using 'statemachine' as a template</h3>
\ No newline at end of file
+noMain: true
+---
```
New line here as well

montymxb commented on hugo/content/showcase/statemachine.html:
```diff
@@ -5,100 +5,9 @@
-    window.addEventListener("resize", () => client.updateLayout());
-</script>
\ No newline at end of file
+---
```
New line

montymxb commented on package.json:
```diff
@@ -11,19 +11,12 @@
     "hugo",
     "tailwind"
-  ],
-  "volta": {
```
I think we want to keep the volta config if we can. That has helped a lot for isolating env related issues.

montymxb reviewed (commented):
Nice, this will make a really nice addition 🙂 .

Some feedback:
I would consider regrouping some of those entries under hugo/assets/scripts to separate the langium (ast) releated helpers & the language specific logic. A suggestion might look like this in **hugo/assets/scripts/**:
- langium-utils/
- arithmetics/
- statemachine/
- domainmodel/

Others can probably be left flat in the scripts folder (babel & config.toml). But this will make it easier to find all language-specific scripts in one spot. Then if we want to work in something like **lox** or **minilogo**, it will be pretty clear how to add those in.

As for the rebase, I would be happy to take a stab at it to get things up to date (and playing nice with the playgrounds as well).

montymxb commented:
I've added up a branch [lpd-2](https://github.com/langium/langium-website/tree/lpd-2) that rebases this PR on main, and fixes up some remaining compatibility issues as well. Once this looks good (checking with Kai & @emilkrebs ), we can force push the new version here for further review.

msujew commented on hugo/package.json:
```diff
@@ -25,17 +24,29 @@
+    "@babel/preset-typescript": "~7.18.6",
+    "@typefox/monaco-editor-react": "1.0.0-next.6",
+    "hugo-extended": "~0.89.4",
+    "langium": "1.1.0",
+    "langium-arithmetics-dsl": "~1.0.0",
+    "langium-domainmodel-dsl": "~1.0.0",
+    "langium-statemachine-dsl": "~1.0.0",
```
The showcase doesn't run. I assume this is the culprit? These versions should all be the same.

emilkrebs replied:
Only statemachine is implemented yet. Does statemachine work for you?

msujew replied:
The showcase looks like this:

![image](https://user-images.githubusercontent.com/4377073/224825287-91fa6ca8-66ed-4725-bd9b-012e2ce5da31.png)

Note that between 1.0.0 and 1.1.0 of Langium, the framework contains breaking changes which likely lead to runtime errors. If you're using langium `1.1.0`, you should use the same version for `langium-statemachine-dsl`.

montymxb replied:
I'm pretty sure that's a tailwinds thing, I'm certain I saw this before as well, and thought it was fixed in here already 🤔 . I believe Monaco & everything is there, but the styling makes it effectively invisible, otherwise the states would never have populated in the first place.

emilkrebs replied:
Fixed the Tailwind config file, just a one-liner.

montymxb reviewed (approved):
:octocat: Looks good to me. @msujew want to double check that the result is now working on your end before we take this?
